ktor-http/ktor-http-cio/src/io/ktor/http/cio/internals/WeakTimeoutQueue.kt

```diff
@@ -142,15 +145,48 @@ class WeakTimeoutQueue(
         }
     }
 
-    @UseExperimental(InternalCoroutinesApi::class)
-    private class WeakTimeoutCoroutine<in T>(context: CoroutineContext, val delegate: Continuation<T>) :
-        AbstractCoroutine<T>(context, true), Continuation<T> {
-        override fun onCompleted(value: T) {
-            delegate.resume(value)
+    private class WeakTimeoutCoroutine<in T>(
+        context: CoroutineContext,
+        delegate: Continuation<T>,
+        val job: Job = Job(context[Job])
+    ) : Continuation<T>, Job by job, CoroutineScope {
+        override val context: CoroutineContext = context + job
+        override val coroutineContext: CoroutineContext get() = context
+
+        private val state = atomic<Continuation<T>?>(delegate)
+
+        init {
+            context[Job]?.let { parent ->
+                @UseExperimental(InternalCoroutinesApi::class)
+                parent.invokeOnCompletion(onCancelling = true) {
+                    if (it != null) {
+                        resumeWithException(it)
```

Comment:
You are invoking `resumeWithException` twice.
Once here and once in `job.invokeOnCompletion` below